### PR TITLE
doc asm feature - Added new 'Label' section with example and explanations

### LINKED ISSUE #76704: unstable book: explaining how to properly use labels in inline asm
We need a section explaining how to properly use labels in inline asm (i.e. only GAS local labels are allowed).

_Originally posted in https://github.com/rust-lang/rust/pull/76675#issuecomment-691860664_

### PR DESCRIPTION
**Pull request by tdelabro**
Fixes #76704

### CODE REVIEW

File: src/doc/unstable-book/src/library-features/asm.md [hunk 1]
```diff
@@ -372,6 +372,43 @@ unsafe {
 # }
 ```
 
+## Labels
+
+The compiler is allowed to instantiate multiple copies an `asm!` block, for example when the function containing it is inlined in multiple places. As a consequence, you should only use GNU assembler [local labels] inside inline assembly code. Defining symbols in assembly code may lead to assembler and/or linker errors due to duplicate symbol definitions.
+
+Moreover due to [a llvm bug], you cannot use `0` or `1` as labels. Therefore only labels in the `2`-`99` range are allowed.
```
Comment (Amanieu) [inline src/doc/unstable-book/src/library-features/asm.md:379]: It's not just 0 and 1. It's actually anything that can be parsed as a binary literal, so any number that contains only 0 and 1 digits.

Comment (Amanieu) [inline src/doc/unstable-book/src/library-features/asm.md:379]:
```suggestion
Moreover due to [an llvm bug], you cannot use `0` or `1` as labels. Therefore only labels in the `2`-`99` range are allowed.
```

File: src/doc/unstable-book/src/library-features/asm.md [hunk 2]
```diff
@@ -372,6 +372,43 @@ unsafe {
...
+This will decrement the `{0}` register value from 10 to 3, then add 2 and store it in `a`.
+
+This example show a few thing:
+
+First that the same number can be used as a label multiple times in the same inline block.
+
+Second, that when a numeric label is used as a reference (as an instruction operand, for example), the suffixes b (“backward”) or f (“forward”) should be added to the numeric label. It will then refer to the nearest label defined by this number in this direction.
+
+[local labels]: https://sourceware.org/binutils/docs/as/Symbol-Names.html#Local-Labels
+[a llvm bug]: https://bugs.llvm.org/show_bug.cgi?id=36144
```
Comment (Amanieu) [inline src/doc/unstable-book/src/library-features/asm.md:410]:
```suggestion
[an llvm bug]: https://bugs.llvm.org/show_bug.cgi?id=36144
```

Comment (tdelabro) [reply src/doc/unstable-book/src/library-features/asm.md]: done

File: src/doc/unstable-book/src/library-features/asm.md (hunk 1 above)
Comment (tdelabro) [reply src/doc/unstable-book/src/library-features/asm.md]: done

Comment (tdelabro) [reply src/doc/unstable-book/src/library-features/asm.md]: Indeed. I made it clear.

**Comment by tdelabro** [comment]
@Amanieu 

https://github.com/rust-lang/rust/pull/84015/checks?check_run_id=2305123299

It seems to me that the `asm!` block are converted to the AT&T asm syntax, while we are using the intel one all along the documentation, making it unable to recognise the used symbols.

**Comment by Amanieu** [comment]
Ouch, I thought that bug was fixed in LLVM 10. It seems that it is only fixed in 10.0.1, which our CI doesn't have. So we may have to add `allow_fail` back to make CI pass.

**Comment by Amanieu** [comment]
Can you run the examples manually on the playground to ensure they work with the nightly compiler?

**Comment by Amanieu** [comment]
Note that the issue here is that LLVM is incorrectly adding `%` to register names even though we have told it to use Intel syntax. This is fixed in LLVM 10.0.1.

**Comment by tdelabro** [comment]
So, I build the documentation locally, opened it in my browser and ran each example by clicking on the play arrow. They all worked well.

I have added the `allow_fail` flags back. It should pass the CI now

**Comment by tdelabro** [comment]
This is my first MR to the rust repo so I'm unsure. Is there something I'm expected to do before we merge this PR ?

**Comment by Amanieu** [comment]
> This is my first PR to the rust repo so I'm unsure: is there something I'm expected to do before we merge this PR ?

Nope! Once the pull request is approved with r+, bors will automatically merge it after running tests.

**Comment by tdelabro** [comment]
Nice ! Thanks for you help and review !

**Pull request merged**